lib/adapters/command-adapter.ts

```diff
@@ -0,0 +1,116 @@
+import {LanguageClientConnection} from '../languageclient';
+import {ServerCapabilities} from 'vscode-languageserver-protocol';
+import {DisposableLike} from 'atom';
+import * as UUID from 'uuid/v4';
+
+const GLOBAL: any = global;
+
+export class CommandAdapter implements DisposableLike {
+
+    private registrations: Map<string, string[]>;
+
+    constructor(private connection: LanguageClientConnection) {
+        this.registrations = new Map();
+        connection.onRegisterCommand(registration => {
+            if (registration.registerOptions && Array.isArray(registration.registerOptions.commands)) {
+                this.registerCommands(registration.id, registration.registerOptions.commands);
+            }
+        });
+        connection.onUnregisterCommand(unregisteration => this.unregisterCommands(unregisteration.id));
+    }
+
+    initialize(capabilities: ServerCapabilities) {
+        if (capabilities.executeCommandProvider && Array.isArray(capabilities.executeCommandProvider.commands)) {
+            this.registerCommands(UUID(), capabilities.executeCommandProvider.commands)
+        }
+    }
+
+    registerCommands(id: string, commands: string[]): void{
+        const cmdRegistry = this.getLspCommandRegistry();
+        const registeredCommands = commands.filter(cmd => {
+           const handler = (params: any[]) => this.connection.executeCommand({
+               command: cmd,
+               arguments: params
+           });
+           if (cmdRegistry.register(cmd, handler)) {
+               return true;
+           } else {
+               console.error(`Trying to register duplicate command: "${cmd}"`)
+           }
+        });
+        if (this.registrations.has(id)) {
+            throw new Error(`Duplicate registration id: ${id}`);
+        }
+        this.registrations.set(id, registeredCommands);
+    }
+
+    executeCommand(id: string, params: any[]): Promise<any> {
+        return this.getLspCommandRegistry().execute(id, params);
+    }
+
+    unregisterCommands(id: string) {
+        if (this.registrations.has(id)) {
+            const commands = this.registrations.get(id);
+            const cmdRegistry = this.getLspCommandRegistry();
+            if (commands && Array.isArray(commands)) {
+                commands.forEach(command => cmdRegistry.unregister(command));
+            }
+            this.registrations.delete(id);
+        }
+    }
+
+    dispose() {
+        const cmdRegistry = this.getLspCommandRegistry();
+        this.registrations.forEach(commands => commands.forEach(command => cmdRegistry.unregister(command)));
+        this.registrations.clear();
+    }
+
+    private getLspCommandRegistry(): LspCommandRegistry {
+        if (!GLOBAL.lspCommandRegistry) {
```

Comment:
Shouldn't this be per-server?  Is there an advantage to making it available globally to all instances of atom-languageclient?

Reply:
There is an advantage to it. If this registry is global then it is possible to make language servers communicate between each other. The communication to JDT LS is great for the case of Spring Tools because it's mostly an add-on over the Java tooling. In formation spring tools may want from JDT LS is classpath, search, javadoc etc. 

Example
I have extended LSP with `spring/javadoc` request message with java artifact binding key and project uri as parameters for the message.
The message is received on my Atom language client extension.
Access global command registry and look for command with id `jdt/javadoc` for example which would the command registered by JDT LS (`ide-java` package)
Execute the command and pass the binding key and project URI as parameters to the `jdt/javadoc` command
The command execution goes to the JDT LS and comes back with the result via LSP messages
The javadoc content is extracted from the result of the command execution and sent as a reply to the initial `spring/javadoc` command
Thus, I have javadoc in my spring tools LS from JDT LS at the end of this

There is a related PR for the `ide-java` package that uses the global registry: https://github.com/atom/ide-java/pull/79

Nevertheless, I understand your concern with the global registry... It also introduces new API to work with the registry... I wish there was something in Atom core that could be utilized for this purpose.